File: scripts/check-i18n.mjs
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const messagesDir = path.join(root, "src/i18n/messages");
const locales = ["en", "es", "pt"];

const loadKeys = (locale) => {
  const filePath = path.join(messagesDir, `${locale}.json`);

  if (!fs.existsSync(filePath)) {
    console.error(`[i18n] Missing message file: ${filePath}`);
    process.exit(1);
  }

  const data = JSON.parse(fs.readFileSync(filePath, "utf8"));

  if (typeof data !== "object" || data === null || Array.isArray(data)) {
    console.error(`[i18n] ${filePath} must be a flat JSON object.`);
    process.exit(1);
  }

  return Object.keys(data).sort();
};

const keySets = Object.fromEntries(
  locales.map((locale) => [locale, loadKeys(locale)])
);
const referenceLocale = "en";
const referenceKeys = new Set(keySets[referenceLocale]);
let failed = false;

for (const locale of locales) {
  if (locale === referenceLocale) {
    continue;
  }

  const keys = new Set(keySets[locale]);
  const missing = [...referenceKeys].filter((key) => !keys.has(key));
  const extra = [...keys].filter((key) => !referenceKeys.has(key));

  if (missing.length > 0) {
    failed = true;
    console.error(`\n[i18n] ${locale} is missing ${missing.length} key(s):`);
    for (const key of missing) {
      console.error(`  - ${key}`);
    }
  }

  if (extra.length > 0) {
    failed = true;
    console.error(`\n[i18n] ${locale} has ${extra.length} extra key(s):`);
    for (const key of extra) {
      console.error(`  - ${key}`);
    }
  }
}

if (failed) {
  process.exit(1);
}

console.log(
  `[i18n] ${referenceKeys.size} keys × ${locales.length} locales OK (${locales.join(", ")}).`
);

File: scripts/check-i18n.py
#!/usr/bin/env python3
"""Validate that all locale message files share the same keys."""

from __future__ import annotations

import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
MESSAGES_DIR = ROOT / "src" / "i18n" / "messages"
LOCALES = ("en", "es", "pt")
REFERENCE_LOCALE = "en"


def load_keys(locale: str) -> list[str]:
    file_path = MESSAGES_DIR / f"{locale}.json"
    if not file_path.is_file():
        print(f"[i18n] Missing message file: {file_path}", file=sys.stderr)
        raise SystemExit(1)

    with file_path.open(encoding="utf-8") as handle:
        data = json.load(handle)

    if not isinstance(data, dict):
        print(f"[i18n] {file_path} must be a flat JSON object.", file=sys.stderr)
        raise SystemExit(1)

    return sorted(data.keys())


def main() -> int:
    key_sets = {locale: set(load_keys(locale)) for locale in LOCALES}
    reference_keys = key_sets[REFERENCE_LOCALE]
    failed = False

    for locale in LOCALES:
        if locale == REFERENCE_LOCALE:
            continue

        keys = key_sets[locale]
        missing = sorted(reference_keys - keys)
        extra = sorted(keys - reference_keys)

        if missing:
            failed = True
            print(f"\n[i18n] {locale} is missing {len(missing)} key(s):", file=sys.stderr)
            for key in missing:
                print(f"  - {key}", file=sys.stderr)

        if extra:
            failed = True
            print(f"\n[i18n] {locale} has {len(extra)} extra key(s):", file=sys.stderr)
            for key in extra:
                print(f"  - {key}", file=sys.stderr)

    if failed:
        return 1

    print(
        f"[i18n] {len(reference_keys)} keys × {len(LOCALES)} locales OK "
        f"({', '.join(LOCALES)})."
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
